From d4d2aa6d2ef713e36c6f2ec1b20866713783d250 Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Wed, 28 Nov 2018 23:04:43 +0200
Subject: [PATCH 0820/1051] net: mvpp2: recycle extra check for
 zero-copy-buffer in skb

The skb could have special Zero-Copy buffer attached to this skb.
This skb is non-recyclable.
Add checker of shinfo->tx_flag for bit SKBTX_DEV_ZEROCOPY.

Change-Id: Id1f9e49e40e1bf17e7755b66694d9240abe8edc7
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/61423
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
Tested-by: Stefan Chulski <stefanc@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index dbe462b440ec..6e46a23e4bb9 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -3267,6 +3267,9 @@ static int mvpp2_recycle_get_bm_id(struct sk_buff *skb)
 	/* Use skb->cloned but not skb_cloned(), skb_header_cloned() */
 	if (skb_shared(skb) || skb->cloned)
 		return -1;
+	if (skb_shinfo(skb)->tx_flags & SKBTX_DEV_ZEROCOPY)
+		return -1;
+
 	/* Get bm-pool-id */
 	hash &= MVPP2_RXTX_HASH_BMID_MASK;
 	if (hash >= MVPP2_BM_POOLS_NUM)
-- 
2.17.1

